### DISCUSSION TITLE
On Scenery Objects

### DISCUSSION THREAD

**Discussion by tajmone** [alan-if/alan-i18n#34] [Library Design]
I've noticed that the implementation of `scenery` objects is somewhat limited.
Here's the entire source module:

```alan
-- "scenery.i"

Every scenery IsA object
  Is
    Not searchable.
    Not takeable.
    Not pushable.
    Not touchable.

  Verb examine, take
    Does only
      "That's only scenery."
  End verb.
End every.

```
## Honor `xDesc` Attribute

The main problem I see with scenery objects is that in order to produce some text when examining them you need to implement the `Verb examine Does only` on the specific instance, whereas it would be much easier to separate the `examine` verb definition above and tweak it to honor an `xDesc` attribute which is not an empty string (like we did for examine and elsewhere).

It makes sense to me that many scenery objects are usually intended to be examinable, since they are the to provide mood and atmosphere to the game.

## Improve Default Message

Also, I'm not quite happy with the "That's only scenery." default response.

1. It should check if the object is plural ("These are...").
2. I'm not quite convinced that word "scenery" is appropriate here:
    + First of all, this terms is mostly known to IF veterans, and a first time player might be puzzled by it.
    + It kind of breaks the narrative illusion, as if remind us that we're in an artificial set, where plants are just fake plastic props, and horizons are just painted backdrops on some wood panel. There must be a better way to deliver the sense that it's not an important item for the game goals, without breaking the illusion of being in a "real narrative world".

I think the goal of the default message is to inform the player that he/she is dealing with an object that plays no significant role in the game (except maybe revealing some info or hinting in the right direction). So, trying to take it, push it, etc., should result in a response that makes it clear that it's not worth attempting further actions because the object has a merely scenic purpose (without being so explicit though).

- [ ] Any suggestions as to a good default reponse?

## Blocked Attributes vs Verbs Overrides

For the same reasons mentioned above, I can't but notice that scenery objects are trying to prevent actions via attributes negations:

```alan
Every scenery IsA object
  Is
    Not searchable.
    Not takeable.
    Not pushable.
    Not touchable.
```

Although setting those attributes is probably a good practice, I still think that the best approach is overriding as many verbs as possible, to ensure the default (deterrent) response is shown, otherwise these attributes will prevent specific actions without clarifying that attempting other verbs is futile — which might lead to player frustration, especially if he/she is stuck in the game and is going for the "guess the verb" brute force solution to move in the adventure.

Since many verbs share similar syntaxes, it should be possible to override all library verbs with a limited number of verb overrides. The real problem might be that the negated attributes might be triggering CHECKs that will prevent the verbs execution, so overriding the verbs might not work, unless we set all those attribute to true, confiding that our custom `DOES ONLY` verbs on the class will prevail. Authors should still be able to override via custom `DOES ONLY` verbs on specific instances (if the need be).

@thoni56:

- [ ] do you foresee any problems with either of the above approaches?
- [ ] Do you advise one over the other?

Also, the attributes list needs to be checked for it might be incomplete — e.g. I'd add `Not wearable`, `Not eatable`, etc., even though these are unset by default, just to be on the safe side.

**Comment by thoni56** [comment]
I think that the "scenery" implementation is modeled after "That's just scenery." that you find in some games. Things that are there just for ... scenery, not playing any part in the game, not even as a prop. With it, you can go through your location descriptions and create scenery objects for all the nouns. In this way, the player doesn't get "I don't understand the word 'painting'" when the game has just said there was a painting there.

If you have long descriptive texts which mention many nouns, this needs to be as simple as remotely possible.

I feel there is a stylistic variance between "That is just scenery." (that you might want to keep) and a more "illusion preserving" response. 

Maybe there is room for something slightly different. Perhaps a type of thing that is fairly easy to create and works as you would expect with descriptions and all necessary verbs.

I just wanted to comment on this, before proceeding.

Expanding on the scenery class to provide all that "protection" is probably a good idea. And, if done inside that class it should not present any further burden on the author. So I think this is a point of possible improvement.

## Default Response

Perhaps just expand on the current to

    Say The this. "is just part of the scenery."

That fixes issues with forms for the item as such. Still have the "is" though, but that's an easy extension of the suggestion provided that the author can indicate the plural-ness, which shouldn't be an issue:

    The plants Isa scenery Is plural. End The plants.

But also, when it comes to "scenery" players might probably disregard any grammatical errors, so I'd say fixing that is only necessary for completeness and perfection ;-)

(Other languages might require other aspects to resolve this, I suppose.)

## Blocked Attributes vs Verbs Overrides

It seems like overriding verbs is the safe route to go. Then the default response can be given in unconditional checks for all blocked verbs:

    Verb take
        Check <default_response>
    End.

Since checks are always run first starting with the most specialized, this will override any general checks, for objects e.g. And then we don't need the negated attributes.

You are probably writing more Alan code than me these days so, I might have missed something.

**Reply by tajmone** [reply]
> You are probably writing more Alan code than me these days so, I might have missed something.

In the `dev_alan-en` branch I've updating both the code and the documentation, but not much more than mentioned above, and was waiting for some feedback like this before making those changes "final" and merge them into `main` as a new update.

# Default Response

> I think that the "scenery" implementation is modeled after "That's just scenery." that you find in some games.

From what I can recall, most adventures tend to mention something along the lines of "That's not important for the game.", but that might differ from one IF system to the other — in Italian adventures, that's definitely the approach most commonly used.

Right now, I've changed the message to "Forget the OBJECT, [it's/are] not important." — i.e. a clear deterrent to attempt further futile actions on that object.

But now if the scenery object as an `xDesc` attribute it will be honoured and printed when examined, instead of the default message — after all, scenery objects are there to build up the game mood, and a detailed 'examine' description can contribute in that direction.

> I feel there is a stylistic variance between "That is just scenery." (that you might want to keep) and a more "illusion preserving" response.
>
> Maybe there is room for something slightly different. Perhaps a type of thing that is fairly easy to create and works as you would expect with descriptions and all necessary verbs.

I personally think that preserving the narrational illusion is paramount, and although I do understand that "is just scenery" could be inferred to mean "part of the surroundings", I still think there's something unnatural about that response — which brings the ever pending question of "Who does the narrator voice belong to?"

Of course, you can't get rid of the narrator, especially in IF games where there needs to be an extra-diegetic voice reporting parsing errors, asking for disambiguation, etc. But if I try to 'look behind' a scenery object (not knowing it's one) and I get the "The OBJECT is just scenery" response, I can't avoid feeling that the narrator is remind me that thinks around me are not real, and that I should lower my expectations, whereas if it tells me "Forget the OBJECT, it's not important." I feel more that this voice is guiding me through my adventure, protecting me from futile actions and keeping me on the right path, so to speak.

Usually one would resort to scenery to cover details in descriptions, i.e. what we call "implementing what's mentioned", whereas more important unreachable objects should be implemented somewhat more specifically (since they are neither "just scenery" nor "not important"), e.g.

```
Forest Brink

As you emerge from the cursed forest you find yourself in an open plain, cut in
half by a large river. Across the river you can finally see Mount Olympus,
where you final mission of defeating Zeus awaits you.

> x mount olympus

Far away on the horizon, Mount Olympus...(etc.).

> through stone at mount olympus

It's too far away!
```

That's a classic example of scenery objects usage, but here a response like "Mount Olympus it's just scenery" would sound as if the adventure engine were telling you clearly and plainly that the mount on the horizon is just painted on some backdrop board (which is true from the implementation perspective, but definitely wrong from the narrative perspective). Replying "Mount Olympus is not important" would be equally wrong here, for reaching this location is a main goal in the adventure.

I wonder if similar examples should be hand-implemented (without using `scenery`) in order to allow custom responses like "It's too far away!", or whether we should make the default scenery response customizable — i.e. by adding a dedicated `action_response` string attribute, which could replace the default message if not empty. This would allow keeping the default response in most cases, but also override it, just like we did for `xDesc` and 'examine', but whereas 'examine' is a special case that justifies it, I'm not sure whether this extra attribute would be overdoing it, and if leaving such edge cases to be manually implemented would be best, and (most important) more inline with the minimalistic "foundational" nature of library — handling defaults and their overriding is always a tricky balance between keeping things simple and making the library flexible to use.

Ultimately, there's really no kind way to go about telling the player that some objects are just empty props on a stage, but I believe that focusing on their lack of importance is more kind, for it focuses on their role in the overall story, rather than stressing (or betraying) their hollow scenic nature.

Probably there's a more elegant sentence that "is not important", but I couldn't come up with one right now. Surely, we can amend it when we find a better one, but I still think that my current proposal should work better than the "is scenery" approach.

Of course, end author could always teak the default response by editing the library sources.


> Perhaps just expand on the current to
>
>     Say The this. "is just part of the scenery."
>
> That fixes issues with forms for the item as such. Still have the "is" though, but that's an easy extension of the suggestion provided that the author can indicate the plural-ness, which shouldn't be an issue:

Handling number is now fixed (in dev branch), so that's not an issue, as long as the author defines it as `plural` when needed.

The response "The OBJ is just part of the scenery." is definitely more elegant than the previous "is just scenery" (it's that _just_ word which belittles the object), but unfortunately it would only work for things like the sun, the sky, and other natural things of the surroundings, but not quite as well will nearby man-made objects. In IF games, anything can be a scenery, but it might not necessarily qualify as "part of the scenery" — imagine a horror game were there piles of corpses, and they are mentioned as being "part of the scenery", it's odd. I think that focusing on their purpose for the game goals is less risky and more justified.

>  With it, you can go through your location descriptions and create scenery objects for all the nouns. In this way, the player doesn't get "I don't understand the word 'painting'" when the game has just said there was a painting there.

I've mentioned this in the updated documentation, and how to implement them sceneries as a mean to avoid the "I don't understand" response, also mentioning that multiple parts of a description should be just made `synonyms` of the main scenery (it's OK to show the same message, the player understands this, since it's conventional enough in IF games).

# Blocked Attributes vs Verbs Overrides

>It seems like overriding verbs is the safe route to go. Then the default response can be given in unconditional checks for all blocked verbs:
>
>     Verb take
>         Check <default_response>
>     End.

I'll try that. I hadn't thought about it.

I believe that unsetting all the various attributes (`takeable`, `pushable`, etc.) should still be done, since these are also used by other verbs (e.g. 'through OBJ1 at OBJ2') so it's safer to explicitly block any potential action (even by custom author verbs) that might execute if we kept the default values.

**Reply by thoni56** [reply]
> # Default Response
> > I think that the "scenery" implementation is modeled after "That's just scenery." that you find in some games.
> 
> From what I can recall, most adventures tend to mention something along the lines of "That's not important for the game.", but that might differ from one IF system to the other — in Italian adventures, that's definitely the approach most commonly used.

I did say _some_ games. I think it's dependent on the particular game too.

> Right now, I've changed the message to "Forget the OBJECT, [it's/are] not important." — i.e. a clear deterrent to attempt further futile actions on that object.

Fair enough. Perhaps "You don't need to think about ..." is slightly better than "Forget..." Or even "You don't need to bother with..."
 
> I personally think that preserving the narrational illusion is paramount, and although I do understand that "is just scenery" could be inferred to mean "part of the surroundings", I still think there's something unnatural about that response — which brings the ever pending question of "Who does the narrator voice belong to?"

Aah. I think about "scenery" as in the theatrical sense, as a part of the decor, while you seem to think about it as in the surrounding scenery. So maybe that ambiguity is both explaining our slightly different viewpoints, but also that it probably should be avoided in the default message.

> Usually one would resort to scenery to cover details in descriptions, i.e. what we call "implementing what's mentioned", whereas more important unreachable objects should be implemented somewhat more specifically (since they are neither "just scenery" nor "not important"), e.g.
> 
> ```
> Forest Brink
> 
> As you emerge from the cursed forest you find yourself in an open plain, cut in
> half by a large river. Across the river you can finally see Mount Olympus,
> where you final mission of defeating Zeus awaits you.
> 
> > x mount olympus
> 
> Far away on the horizon, Mount Olympus...(etc.).
> 
> > through stone at mount olympus
> 
> It's too far away!
> ```
> 
> That's a classic example of scenery objects usage, but here a response like "Mount Olympus it's just scenery" would sound as if the adventure engine were telling you clearly and plainly that the mount on the horizon is just painted on some backdrop board (which is true from the implementation perspective, but definitely wrong from the narrative perspective). Replying "Mount Olympus is not important" would be equally wrong here, for reaching this location is a main goal in the adventure.

But that is something completely different from what I mean "scenery" objects are. Not everything that is unreachable are, or should be, "scenery". If it is important it should be handled as a normal, important, object. So this data point confuses the issue.

> I wonder if similar examples should be hand-implemented (without using `scenery`) in order to allow custom responses like "It's too far away!", or whether we should make the default scenery response customizable — i.e. by adding a dedicated `action_response` string attribute, which could replace the default message if not empty. This would allow keeping the default response in most cases, but also override it, just like we did for `xDesc` and 'examine', but whereas 'examine' is a special case that justifies it, I'm not sure whether this extra attribute would be overdoing it, and if leaving such edge cases to be manually implemented would be best, and (most important) more inline with the minimalistic "foundational" nature of library — handling defaults and their overriding is always a tricky balance between keeping things simple and making the library flexible to use.

Exactly! So it is important to describe what a "scenery" object is and what is should and should not be used for.
 
> Ultimately, there's really no kind way to go about telling the player that some objects are just empty props on a stage, but I believe that focusing on their lack of importance is more kind, for it focuses on their role in the overall story, rather than stressing (or betraying) their hollow scenic nature.

True. Unless you want to make that a part of the tone you want in your game.

> Probably there's a more elegant sentence that "is not important", but I couldn't come up with one right now. Surely, we can amend it when we find a better one, but I still think that my current proposal should work better than the "is scenery" approach.
> 
> Of course, end author could always teak the default response by editing the library sources.
> 
> > Perhaps just expand on the current to
> > ```
> > Say The this. "is just part of the scenery."
> > ```
> >   
> > That fixes issues with forms for the item as such. Still have the "is" though, but that's an easy extension of the suggestion provided that the author can indicate the plural-ness, which shouldn't be an issue:
> 
> Handling number is now fixed (in dev branch), so that's not an issue, as long as the author defines it as `plural` when needed.
> 
> The response "The OBJ is just part of the scenery." is definitely more elegant than the previous "is just scenery" (it's that _just_ word which belittles the object), but unfortunately it would only work for things like the sun, the sky, and other natural things of the surroundings, but not quite as well will nearby man-made objects. In IF games, anything can be a scenery, but it might not necessarily qualify as "part of the scenery" — imagine a horror game were there piles of corpses, and they are mentioned as being "part of the scenery", it's odd. I think that focusing on their purpose for the game goals is less risky and more justified.

Again, it is always a matter of how much work an author want to put into crafting each and every object (which you can) versus how much breaking the illusion (s)he think that is worth. All IF is a very thin layer on top of a set of canned messages, it is not even a simulation.

> 
> > With it, you can go through your location descriptions and create scenery objects for all the nouns. In this way, the player doesn't get "I don't understand the word 'painting'" when the game has just said there was a painting there.
> 
> I've mentioned this in the updated documentation, and how to implement them sceneries as a mean to avoid the "I don't understand" response, also mentioning that multiple parts of a description should be just made `synonyms` of the main scenery (it's OK to show the same message, the player understands this, since it's conventional enough in IF games).

Good.
 
> # Blocked Attributes vs Verbs Overrides
> > It seems like overriding verbs is the safe route to go. Then the default response can be given in unconditional checks for all blocked verbs:
> > ```
> > Verb take
> >     Check <default_response>
> > End.
> > ```
> 
> I'll try that. I hadn't thought about it.
> 
> I believe that unsetting all the various attributes (`takeable`, `pushable`, etc.) should still be done, since these are also used by other verbs (e.g. 'through OBJ1 at OBJ2') so it's safer to explicitly block any potential action (even by custom author verbs) that might execute if we kept the default values.

If unconditional checks work out I think negating all attributes would be duplication since all such attribute tests are performed in checks too. So if possible I think they should be avoided. It also lessens the confusion about what is the priority between the attributes and the unconditional checks. Unless you have an example where the attribute setting would make a difference?

What is not entirely clear to me is how an author would handle "scenery" objects w.r.t. any new verbs that they might create. Subclassing the scenery class and then adding an unconditionally checked verb implementation in that class for the added verbs would probably be the right way to do it. And of course, remember to use "Is my_scenery" instead of "Is scenery"...

**Reply by tajmone** [reply]
## Blocked Attributes vs Verbs Overrides

> Since checks are always run first starting with the most specialized, this will override any general checks, for objects e.g.

I've tried to override 'take' on `scenery` using an unconditional Check:

```alan
Every scenery IsA object
  ...
  Verb take
    Check
      Say "Forget $+1,".
      If this is not plural
        then "it's"
        else "they're"
      End if.
      "not important."
  End verb.
```

but it doesn't work as expected, I'm still getting the message from the `is takeable` check from 'take' on `object`. So, it seems like the order of CHECKS execution is the other way round (from the generic down to the specialized). The chapter _[Verb Execution]_ says:

> The table above illustrates the normal order of execution of verb bodies and checks. Starting with any base classes to the outermost region (containing location), continuing to the actual instance of that location...

IRC, it's the execution of verb checks is outward-to-inward, like plain `Does` verbs, and it's only the `Does Only`/`Before`/`After` verbs that have the inward-to-outward execution.

Therefore, we can't bypass the default message because:

- A `Does Only` definition won't work, since CHECKs are executed generic to specific, so the verb body never gets executed.
- A `Check` (conditional or unconditional) won't work because of the same generic precedence problem.

which leads to the problem of having to unset attributes...

> If unconditional checks work out I think negating all attributes would be duplication since all such attribute tests are performed in checks too.

I'm afraid some verbs are checking for the attributes in the verb body instead. I'm not entirely sure why its so, but I remember seeing this in the StdLib too, so my guess is that it has to do either with the fact that some of these attributes are only available on `object` (preventing checks for verbs which are expected to also be usable with actors, albeit failing with some polite message), or due to some execution order precedence, or the desire to allow authors to override them via `Does only` (which the checks would block).

> Unless you have an example where the attribute setting would make a difference?

Verbs with implicit taking would be a good example, but we've dropped implicit taking from the library (but authors might still implement it in their custom verbs though). E.g. 'through statue at Bob' (`through (obj1) at (obj2)`). If we add `Check obj1 is takeable` on the verb, then authors won't be able to enforce a custom message on specific instances (same problem as with scenery 'take').

I've stumbled upon these limitations many times over in the StdLib too. The problem is that CHECKs in ALAN are _definitive_ — once added to a high level class, there's no way to bypass them down the line. Add to that the impossibility of a verb body to interrupt the chain of execution (i.e. a specialized class or instance preventing further execution of generic `Does` and `Does After` bodies along the line), and you can quickly see that implementing general purpose verbs is not always easy.

So much so that I even though of moving all the StdLib definitions from `actor` and `object` to subclasses `lib_actor` and `lib_object`, in order to leave the original classes "clean" for end authors to override them. That might acceptable in an Italian library, but would look odd on an English library (IMO).

### Anssi's Solution: An All-Encompassing Location

The way Anssi worked around this in the StdLib was to add a location encompassing all other locations (via `Initialize`, where every location is looped through and moved AT this super location), which allows authors to override _any_ verb on that location, which takes precedence since it's always the outermost instance in the game. But this only grants the possibility of overriding classes which have a `Does Only` body, not their CHECKs! Of course, if all you're trying to do is override a message, then this can be done via an additional `Check` on the super location, which is always executed first; but if you need to carry out more complex code operations, that there's no way around it.

This is the reason why you often see attributes checking done in the verb body rather than its `Check` part, because it allows authors to override verbs on instances.


## Default Response

> I did say _some games_. I think it's dependent on the particular game too.

If it's just a matter of the game genre, than authors can simply replace the default message; but if its tied to the specific objects, than exposing a variable to optionally customize the message on per object basis might make sense (a bit overkill, but still...).

> Fair enough. Perhaps "You don't need to think about ..." is slightly better than "Forget..." Or even "You don't need to bother with..."

Indeed, something along those lines (I don't like much the "Forget..." part either, it has to be something more subtle).

> But that is something completely different from what I mean "scenery" objects are. Not everything that is unreachable are, or should be, "scenery". If it is important it should be handled as a normal, important, object. So this data point confuses the issue.

The reason I'm bringing it up is to help us define what are the boundaries and scope of scenery objects — bare in mind that Inform7 also has backdrops objects, which are scenery-like objects but visible in every room (or in specific regions), usually employed to implement the sky, persistent objects on the horizon, etc. In ALAN we'd use nested locations to achieve this, something which should be mentioned in the docs section on the scenery class.



[Verb Execution]: https://alan-if.github.io/alan-docs/manual-alpha/manual.html#_verb_execution

**Reply by tajmone** [reply]
# Thoughts on CHECKs Limitations...

After my last answer on the problem of overriding CHECKs on `scenery`, I think that @AnssiR66's solution of creating an "artificial" location that will contain all author created locations is possibly the best solution, not only for our `scenery` problem but also to grant end user full power of overriding any library defined verbs.

What do you think? Should we add this location to the Foundation library too?

Of course, it's not just a location instance that needs to be added, but a specific class, since authors (including us library maintainers) will need to use `Add to every` to override verbs "globally". It also requires a startup EVENT to move all locations inside this "global location", preserving any nested locations.

In the StdLib this is handled in [`lib_definitions.i`][lib_definitions.i]:

```alan
  INITIALIZE
    SCHEDULE check_score AFTER 0.
    SCHEDULE check_restriction AFTER 0.
    SCHEDULE check_darkness AFTER 0.

  -- ===========================================================================
  -- Nest Every Location into my_game
  -- ===========================================================================

  -- The following code ensures that every location in the adventure (except for
  -- `my_game` itself and the special `nowhere` location) will become nested
  -- into the `my_game` location.

  -- Thanks to this, any verb defined on `my_game` will be globally available,
  -- which is what makes it possible to override a predefined library verb by
  -- re-defining it on the `my_game` instance --- since `my_game` is always the
  -- outermost nester of any location, all verbs defined on it will always be in
  -- scope, anywhere.


    FOR EACH l IsA LOCATION
      DO
        EXCLUDE nowhere FROM nested OF l.
        IF COUNT IsA LOCATION, AT l > 0
          THEN
            FOR EACH x IsA LOCATION, AT l
              DO
                INCLUDE x IN nested OF l.
            END FOR.
        END IF.
    END FOR.

    FOR EACH l IsA LOCATION
      DO
        IF l <> my_game AND l <> nowhere
          THEN LOCATE l AT my_game.
        END IF.
    END FOR.

   [ and more ...]
```

Probably our code will be shorter, since our library doesn't define special location classes (which in the StdLib need special handling), but the principle is more or less the same: iterate al locations, track their nesting level, move the location in the "global location", ensure the original nesting of locations is preserved.

I'm not sure is this is the type of thing which breaks the concept of the library being "foundational", and if it's worth having to adding to the location class that Set-attribute to track nesting, but I can see why @AnssiR66 came up with solution — it's really smart, and it took quite a while to realize how powerful it is in the StdLib (for almost a year working on the StdLib source I didn't even realize it was there, so much its presence is discrete, and only really discovered it while working on the StdLib Manual examples).

Also, I'm not sure if the initialization code might potentially end up clashing with author's custom location classes. From what I remember, most of the code omitted in the above snippet is just code that handles library defined classes, i.e. enforcing specific locations nesting that are required for StdLib specific classes (e.g. ensuring that every ROOM is AT INDOOR, and every SITE at OUTDOOR), but I think that since the initialization code ends up reconstructing the rooms nesting hierarchy correctly, simply moving them inside the global location, it should be a fairly safe operation.

Although it might seem an advanced feature, I think that @AnssiR66 added it to solve the general problem of allowing authors to _always_ bypass any library defined verbs, so it might make sense to adopt it even for the Foundation.

How do you feel about this solution?

[lib_definitions.i]: https://github.com/AnssiR66/AlanStdLib/blob/dev-2.2.0/StdLib/lib_definitions.i#L364

# CHECK AFTER/BEFORE?

On the ALAN language side, I thought that the real problem is that the `After` and `Before` clause are only available for VERBs, i.e. that if we had similar constructs (and possibly even `Only`, if it makes sense) for CHECKs, then we could control their order of execution too.

I'm not sure if this is even possible, but the logic seems to be the same as that of VERB executions after all, except that in this case we'd be controlling the order in which CHECKs are carried out.

I do understand that even the `After` and `Before` clause for VERBs are considered "advanced" in ALAN, i.e. in terms of the simplicity of design goal, and that adding them even for CHECKs might be stretching the language beyond its philosophy. But then, the concept would be the same as that provided for VERBs, so if a user has managed learning them on VERBs it doesn't really add much complexity applying the same concept to CHECKs too.

**Reply by thoni56** [reply]
First, I'm really sorry for working of memory here which obviously is leading us astray. So thank you for being the "concrete" part in these discussions!

Since I read your previous comment my mind has been wandering a lot around this without any clear conclusions.

I'm primarily deep-diving into the reasons for having the general-to-specific ordering of the checks.

Remember that the initial decision on ordering of verb execution was based on some experience from v2, but that only covered global verb definitions, verbs in the location and the object. But the decision for ordering of execution in v3, with classes and inheritance, was taken without much experience, in particular of writing larger, real, games using those features. It was really only based on logic and inference.

## Radical, breaking, thoughts

I might be second-guessing myself too much based on this single example, but I'm considering how we could experiment with the execution order to learn more about what would be the best way.

To me (at this moment, the current "me") the ordering of check execution that I assumed in my proposal feels like the natural order. But I feel I need more anecdotal data, examples, where this would come into play, to do a good reasoning here.

I have also thought about controlling check order specifically, like you suggest. And I also feel that would be complex and stretching things  bit too far. It would be acceptable if there was just a few cases where it would be required, not if it actually was the only way to get inherited checks to work the way you mostly wanted.

## Analogy with constructors

An analogy would be how object oriented languages allow a constructor to invoke, or not, the "parents" constructor. The constructor is always run with the exact class being instantiated. (This would indicate "start with the most specific".)

The that constructor could invoke the constructor of the parent class at will. In my experience this is _usually_ done before running class-specific code. (This would indicate "start with the more general".) But you might actually invoke it at any time, including last. (This would again be analogous to "start with the more specific".)

## What do checks do?

We now have 2 examples of when the current order is an obstacle to "natural" needs. Or maybe 1.5 since StdLib and alan-i18n have the same problem with overriding verbs.

I have only two large-ish game sources on my hands both built on top of StdLib 1.0. Typically they have

```
ADD TO EVERY THING
  VERB attack
	CHECK target IS examinable
		ELSE 
			IF target IS NOT plural
				THEN "That's not"
				ELSE "Those are not"
			END IF.
			"something you can $v."
	AND target <> hero 
		ELSE "It doesn't make sense to $v yourself."
	AND target NOT IN hero
		ELSE "It doesn't make much sense to $v something you're holding."
	AND target NOT IN worn
		ELSE "It doesn't make sense to $v something you're wearing."
	AND CURRENT LOCATION IS lit
		ELSE "It is too dark to see."
	AND target IS reachable
		ELSE SAY THE target. 
			IF target IS NOT plural
				THEN "is"
				ELSE "are"
			END IF. 
			"too far away."
```

My imagination is not working well enough to figure out an example where I'd sub-class THING and implement a more specific check for "attack". But that is really what would be needed to research this line of thought thoroughly.

I guess the question is "what problems might there be with a more specific check triggering before a more general?"

The syntax guarantees that the instance is an instance of the correct class and the compiler ensures that all access to aspects of the parameter is legal given that it is of that class, so no crashes should occur.

So it is only "logical" problems that might occur, such as a very specific message is given when some more general aspect should have already been checked. I think this is the reasoning behind the current implementation.

## Conclusion

None. Yet. I'm just letting you know where I'm in this at this point. Any contributions, examples, thoughts would be most welcome.

I will probably make a test with "inverted" check executions order just to see what it feels like and perhaps try to use that with one of the games.

**Reply by tajmone** [reply]
# General 2 Specific

> Remember that the initial decision on ordering of verb execution was based on some experience from v2, but that only covered global verb definitions, verbs in the location and the object. But the decision for ordering of execution in v3, with classes and inheritance, was taken without much experience, in particular of writing larger, real, games using those features. It was really only based on logic and inference.

I think that the original logic of applying CHECKs from the general to the specific still holds true, and its in line with the `WHERE` clauses, which are a good example of why and where class checking should be follow that pattern.

But it's also true that inheritance has brought into ALAN new possibilities, and along with them new challenges that can lead to complexity where authors can bump into practical limitations.

> I have also thought about controlling check order specifically, like you suggest. And I also feel that would be complex and stretching things bit too far. It would be acceptable if there was just a few cases where it would be required, not if it actually was the only way to get inherited CHECKS to work the way you mostly wanted.

## Verbs with Identical Syntaxes Not Allowed

My impression is that this type of problem tends to surface in large games, and more so in libraries and extensions. The question of how to enforce a radically different VERB response and/or behaviour (the former can be achieved via an unconditional CHECK, the latter requires a DOES ONLY) is _primarily_ tied to the order of executions of VERB CHECKs, but not exclusively. We need to keep into account the way SYNTAXes work too — i.e. whether it's possible or not to define another VERB with same parameters but different ID — but doing so would result in the following compiler error:

    334 W : Multiple syntaxes are defined with this structure (words and parameters).

E.g. if it was possible to define a `take2` SYNTAX and VERB, with identical parameters to `take`, we could then add `VERB take2` to the specific class or instance where we'd like to override the default CHECKs of `take`, since this would be an entirely different VERB where the CHECKs and verb bodies of `take` don't apply. But, as we've seen, this is not a viable option.

Obviously, if there was a `CHECK BEFORE` clause the solution would be very easy and intuitive, since this clause would be used as the exception to the rules, and only when there's a need for the specific to have a better hand on the generic.

## Overriding Verbs on Locations

Having said this, it's not as if authors are left without solutions... It's always possible to exploit the natural generic to specific exec order, especially the fact that enclosing locations have precedence over any instances they contain (and their base classes to, all the way up to `ENTITY`).

So, a simple solution would be to reimplement the VERB on the `location` class. E.g.:

```alan
Add to every location
  Verb take
    Check obj <> fire
      else "Didn't your mum warn you not to play with fire?!"
  End verb.
End Add.
```

In this case, the `take` verb will bypass all the default CHECKs on `object` class, since the current location is always processed first.

### Lack of `Not IsA` Limits

This works fine with specific instances, but if we were to bypass the CHECKs for a whole class (which indeed is our case with the `scenery` class), we stumble upon the limitations of CHECKs not supporting negative `IsA` expression — i.e. this won't work:

```alan
Add to every location
  Verb take
    Check obj not IsA scenery
      else "That's just part of the scenery."
  End verb.
End Add.
```

The topic of lack of `IsA` negation was recently discussed and added to the ALAN repository Issues as a feature request, but this is a practical example of what makes it problematic to handle this type of use case right now, with the current ALAN syntax.

As far as I know, there's currently not workaround to handle CHECKs overriding at the class level. Even using the solution proposed by @AnssiR66 in the StdLib (i.e. to create an all-encompassing location for overriding library defined verbs) doesn't fix the lack of `IsA` negation — as a matter of fact, all the messages overrides of the specialized StdLib classes are not working, due to this problem (they are there in the code, but the test suite showed that they are never displayed).

# General Considerations

So, as you can see, this problem doesn't exist "in a vacuum", so to speak, but has to be viewed in the overall context of the tools available through the ALAN language. Implementing support for negative `IsA` expression (`not IsA` or `IsA not`) would already make this problem much easier to solve — whether the solution should be considered elegant or not is another matter.

So, what would be more elegant and "idiomatic"? having a `CHECK BEFORE` construct or overriding verbs on locations?

I fell that having `BEFORE`, `AFTER` and `ONLY` clauses for `CHECK` would feel natural, since these are already available for `DOES` — and again, these represent exceptions to the rule, and should only be used when the more conservative approach doesn't work.

Implement verbs on locations seems also a reasonable solution, albeit more "hackish" in nature and (possibly) harder to decipher for someone studying an adventure source (requires more abstraction to visualize). Still, it's a good solution, and authors should be aware that encompassing locations (instances or their classes) have execution precedence over their contained instances (when the generic to specific pattern applies). So, if we could use `not IsA` for class CHECKS, we probably be having this discussion anyway (apart from mentioning changes to the original library code, since these options didn't even exist at the time of its writing).

# Analogy with constructors

> An analogy would be how object oriented languages allow a constructor to invoke, or not, the "parents" constructor. The constructor is always run with the exact class being instantiated. (This would indicate "start with the most specific".)

That's a good analogy to bring into the conversation, and that's exactly the approach taken by many IF engines (e.g. Hugo, and I think also TADS).

Usually if a constructor (or any inherited method, really) is defined on the instance (or subclass) then that's executed in place of the constructor of its ancestor classes. And the instance/subclass can always invoke execution of the parent's constructor via `super` (or some similar keyword).

But I'm not sure how this model overlaps with how ALAN inheritance works for VERBs, especially for their CHECKs, since if I've understood correctly CHECKs inheritance simply executes all the CHECKs encountered along the generic-to-specific path (either that or it first scans them and piles them up somewhere before executing them in reverse order). Since CHECKs don't have their own ID, like methods, I don't see how a specific CHECK definition (in a class or instance) could replace another defined elsewhere.

Also, I'm not sure that adopting the OOP constructors model would make things easier for authors, since that would add more complexity. I think that the current inheritance system is easier to understand (despite its limitations) since its much more linear in nature — and the `-xml` option should produce a structural representation of the adventure which should be easy to read and understand (although the `-xml` option seems to no longer work right now).

**Reply by tajmone** [reply]
# Converting Scenery to Attribute

@thoni56, after having though the issue through, I think that the best solution would be to convert the `scenery` class into an attribute. This is also what @AnssiR66 did in the StdLib 2, where in version 1 it was a class (as it had been in ALAN Lib all along); I believe that Anssi stumbled upon the same limitations and realized that it makes more sense to use an attribute instead.

With a `scenery` class we'd have to re-implement new CHECKs for every verb, on the class (which is what's failing to work); with an attribute, we'd have to add a CHECK to every VERB to prevent the action if the parameter `is scenery` — either way, it doesn't make a big difference in terms of work, not for us maintainers, nor for end users implementing custom verbs. The only possible downside is that every verb will have a longer series of CHECKs, whereas re-implementing the VERBs on the `scenery` class would have slightly cleaner, for it would happen all in the `scenery` module.

When you come to think about it, it doesn't make much sense to implement a whole class for scenery. It's not a specialization of a super class, it's actually the contrary, which is probably why it's giving so many problems.

So, if it's OK with you, I'd like to proceed with the English library, then apply the same changes to the Spanish and Italian libs too.

**Reply by thoni56** [reply]
Sure. Go ahead!

**Reply by tajmone** [reply]
# Scenery Now an Attribute!

Done! in Foundation English v0.3.0 the `scenery` class was converted to an attribute. All verbs were adapted accordingly, and full coverage was provided in the test suite.

For the full details, see [CHANGELOG] and the _Foundation Design_ document (local HTML).


[CHANGELOG]: https://github.com/alan-if/alan-i18n/blob/main/alan_en/Foundation/CHANGELOG.md#v030-20211223
